tests/governance_js.py

```diff
@@ -0,0 +1,101 @@
+def test_proposal_validation(network, args):
+    primary, _ = network.find_nodes()
+    add_member = [action("add_member", cert="", enc_pubk="", member_data={})]
+
+    valid_set_recovery_threshold = [action("set_recovery_threshold", threshold=5)]
+    no_args_set_recovery_threshold = [action("set_recovery_threshold")]
+    bad_arg_set_recovery_threshold = [action("set_recovery_threshold", threshold=5000)]
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold)
+        assert r.status_code == 200, r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold * 2)
+        assert r.status_code == 200, r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", no_args_set_recovery_threshold)
+        assert r.status_code == 400 and r.body.json()["error"]["code"] == "ProposalFailedToValidate", r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", no_args_set_recovery_threshold + bad_arg_set_recovery_threshold)
+        assert r.status_code == 400 and r.body.json()["error"]["code"] == "ProposalFailedToValidate", r.body.text()
+
+    return network
+
+@reqs.description("Test proposal storage")
+def test_proposal_storage(network, args):
+    primary, _ = network.find_nodes()
+    add_member = [action("add_member", cert="", enc_pubk="", member_data={})]
+
+    valid_set_recovery_threshold = [action("set_recovery_threshold", threshold=5)]
+
+    with primary.client(None, "member0") as c:
+        r = c.get(f"/gov/proposals.js/42")
+        assert r.status_code == 404, r.body.text()
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold)
+        assert r.status_code == 200, r.body.text()
+        pid = r.body.json()["proposal_id"]
+        r = c.get(f"/gov/proposals.js/{pid}")
+        assert r.status_code == 200, r.body.text()
+
+    return network
+
+@reqs.description("Test ballot storage")
+def test_ballot_storage(network, args):
+    primary, _ = network.find_nodes()
+    add_member = [action("add_member", cert="", enc_pubk="", member_data={})]
+
+    valid_set_recovery_threshold = [action("set_recovery_threshold", threshold=5)]
+
+    with primary.client(None, "member0") as c:
+        r = c.post("/gov/proposals.js", valid_set_recovery_threshold)
+        assert r.status_code == 200, r.body.text()
+        pid = r.body.json()["proposal_id"]
+
+        vote = {"ballot": "export function vote (proposal, proposer_id, tx) { return true }"}
+        r = c.post(f"/gov/proposals.js/{pid}/votes", vote)
+        assert r.status_code == 200, r.body.text()
+
+        member_id = 0
+        r = c.get(f"/gov/proposals.js/{pid}/votes/{member_id}")
+        assert r.status_code == 200, r.body.text()
+        assert r.body.text() == f'"{vote["ballot"]}"'
```

Comment:
I guess this is wip, should be the same schema as what's used for submission: `{"ballot": ..}`

Reply:
It can be, or it can stay unwrapped, I'm happy either way.

Reply:
I would always wrap, so that you can send more info in the future in extra fields.